x-pack/legacy/plugins/maps/public/angular/map_controller.js

```diff
@@ -124,6 +126,77 @@ app.controller('GisMapController', ($scope, $route, kbnUrl, localStorage, AppSta
     globalState: globalState,
   });
 
+  /* Saved Queries */
+  $scope.showSaveQuery = capabilities.get().maps.saveQuery;
+
+  $scope.$watch(() => capabilities.get().maps.saveQuery, (newCapability) => {
+    $scope.showSaveQuery = newCapability;
+  });
+
+  $scope.onQuerySaved = savedQuery => {
+    $scope.savedQuery = savedQuery;
+  };
+
+  $scope.onSavedQueryUpdated = savedQuery => {
+    $scope.savedQuery = { ...savedQuery };
+  };
+
+  $scope.onClearSavedQuery = () => {
+    delete $scope.savedQuery;
+    delete $state.savedQuery;
+    onQueryChange({
+      filters: [],
+      query: {
+        query: '',
+        language: localStorage.get('kibana.userQueryLanguage')
+      },
+    });
+  };
+
+  function updateStateFromSavedQuery(savedQuery) {
+    if (savedQuery.attributes.timefilter) {
+      if (savedQuery.attributes.timefilter.refreshInterval) {
+        $scope.onRefreshChange({
+          isPaused: savedQuery.attributes.timefilter.refreshInterval.pause,
+          refreshInterval: savedQuery.attributes.timefilter.refreshInterval.value
+        });
+      }
+      onQueryChange({
+        filters: savedQuery.attributes.filters || [],
+        query: savedQuery.attributes.query,
+        time: savedQuery.attributes.timefilter
+      });
+    } else {
+      onQueryChange({
+        filters: savedQuery.attributes.filters || [],
+        query: savedQuery.attributes.query,
+      });
+    }
+  }
+
+  $scope.$watch('savedQuery', (newSavedQuery, oldSavedQuery) => {
+    if (!newSavedQuery) return;
+
+    $state.savedQuery = newSavedQuery.id;
+
+    if (newSavedQuery.id === (oldSavedQuery && oldSavedQuery.id)) {
+      updateStateFromSavedQuery(newSavedQuery);
```

Comment:
Calling `updateStateFromSavedQuery` inside the savedQuery watch results in dispatching `onQueryChange` multiple times. `on-query-submit` callback is called when saved query is saved. That triggers `onQueryChange`. Is there ever an instance where `savedQuery` changes and `on-query-submit`, `on-refresh-change`, or `on-filters-updated` has already been triggered?

Reply:
@nreese after consulting with @Bargs, I [deleted a redundant call](https://github.com/elastic/kibana/pull/44442/commits/4d6a62ce06400da9cd87e654b7e39a7e58b95e48#diff-b43b38ab66cf7c95efa52e037eb4b552L268) to `onQuerySubmit` from the [search bar](https://github.com/elastic/kibana/blob/3deb07f6a580ef899c968308f1aa486b900117d7/src/legacy/core_plugins/data/public/search/search_bar/components/search_bar.tsx#L269-L277), thereby reducing the number of time `onQueryChange` is called. @bargs, please confirm the change is what we decided on as a solution.

Reply:
@nreese If I understand your comment, the trouble also lies with calling `updateStateFromSavedQuery` inside the watchers on $state.savedQuery and $scope.savedQuery. I'm looking into that too.

Reply:
I wouldn't call the line removed in 1b8333f redundant. Things only work now because the watcher on `$scope.savedQuery` runs after the async request for the saved query completes. At that point the old and new saved query IDs are the same so the conditional passes and we call `updateStateFromSavedQuery`. But that's not why we added that conditional in the first place. It's an unintentional and confusing side effect and I don't think we should rely on it.

I was thinking about how we set up these watchers some more and I think we're unnecessarily triggering a request for the saved query from the server when we already have it in hand. That's ultimately causing the unnecessary extra call of `updateStateFromSavedQuery` as well. I didn't think I could explain this idea very well in text, so @TinaHeiligers I submitted a PR against your branch to demonstrate the change I'm suggesting. 

https://github.com/TinaHeiligers/kibana/pull/4/files

Reply:
@Bargs the changes have been implemented and `onQueryChange` now only dispatches once per event.

Reply:
Would you mind applying these changes to Discover, Visualize, and Dashboard as well? `updateStateFromSavedQuery` also gets called twice in those apps, and this should solve it.

Reply:
@Bargs Sure, should those changes be in a separate PR or is it ok to add them to this one?